File: frontend/src/types/report.ts
/**
 * Tipos para reportes de desviacion de gastos.
 *
 * Los montos se reciben como string desde el backend (Decimal serializado).
 * No recalcular desviacion ni porcentaje en frontend; usar los valores del backend.
 */

export type ExpenseVariance = {
  year: number;
  month: number;
  cost_center_id: number;
  cost_center_code: string | null;
  cost_center_name: string | null;
  expense_concept_id: number;
  expense_concept_code: string | null;
  expense_concept_name: string | null;
  planned_amount: string;
  actual_amount: string;
  deviation_amount: string;
  /** Ratio decimal con 4 decimales (ej. "0.2000" = 20 %). Null si planned = 0 y actual > 0. */
  deviation_percentage: string | null;
  status: "SIN PRESUPUESTO" | "SOBRECOSTO" | "AHORRO" | "EN PRESUPUESTO" | string;
};

export type ExpenseVarianceFilters = {
  year?: number;
  month?: number;
  cost_center_id?: number;
  expense_concept_id?: number;
};

/**
 * Resumen de desviacion consolidado por año, mes y centro de costo.
 * No incluye detalle por concepto; para eso usar ExpenseVariance.
 *
 * Los montos se reciben como string desde el backend (Decimal serializado).
 * No recalcular desviacion ni porcentaje en frontend; usar los valores del backend.
 */
export type ExpenseAnalysis = {
  year: number;
  month: number;
  cost_center_id: number;
  cost_center_code: string | null;
  cost_center_name: string | null;
  planned_amount: string;
  actual_amount: string;
  deviation_amount: string;
  /** Ratio decimal con 4 decimales (ej. "0.2000" = 20 %). Null si planned = 0 y actual > 0. */
  deviation_percentage: string | null;
  status: "SIN PRESUPUESTO" | "SOBRECOSTO" | "AHORRO" | "EN PRESUPUESTO" | string;
};

export type ExpenseAnalysisFilters = {
  year?: number;
  month?: number;
  cost_center_id?: number;
};

export type ExpenseKpis = {
  year: number | null;
  month: number | null;
  planned_amount_total: string;
  actual_amount_total: string;
  deviation_amount_total: string;
  deviation_percentage: string | null;
  execution_percentage: string | null;
  status: "SIN PRESUPUESTO" | "SOBRECOSTO" | "AHORRO" | "EN PRESUPUESTO" | string;
};

export type ExpenseKpisFilters = {
  year?: number;
  month?: number;
};

File: backend/app/schemas/report.py
"""Esquemas Pydantic v2 para reportes y consolidaciones."""
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel


class ExpenseVarianceRead(BaseModel):
    """Desviacion monetaria por periodo, centro y concepto.

    Formulas:
        deviation_amount     = actual_amount - planned_amount
        deviation_percentage = deviation_amount / planned_amount  (ratio decimal, no porcentaje x100)

    Casos de deviation_percentage:
    - planned_amount > 0:                   ratio con 4 decimales (ej. 0.2000 = 20 %).
    - planned_amount = 0 y actual = 0:      Decimal("0.0000").
    - planned_amount = 0 y actual > 0:      None  (sin presupuesto definido).

    Valores de status:
    - "SIN PRESUPUESTO": planned_amount = 0 y actual_amount > 0.
    - "SOBRECOSTO":      planned_amount > 0 y deviation_amount > 0.
    - "AHORRO":          planned_amount > 0 y deviation_amount < 0.
    - "EN PRESUPUESTO":  deviation_amount = 0 y no aplica SIN PRESUPUESTO.

    Si no existe planificado para la combinacion: planned_amount = Decimal("0.00").
    Si no existe real para la combinacion:         actual_amount  = Decimal("0.00").
    No se calculan semaforos visuales; esos corresponden a tareas posteriores.
    """

    year: int
    month: int
    cost_center_id: int
    cost_center_code: str | None
    cost_center_name: str | None
    expense_concept_id: int
    expense_concept_code: str | None
    expense_concept_name: str | None
    planned_amount: Decimal
    actual_amount: Decimal
    deviation_amount: Decimal
    deviation_percentage: Decimal | None
    status: str


class ActualExpenseConsolidatedRead(BaseModel):
    """Consolidado de gastos reales agrupado por periodo, centro y concepto.

    actual_amount es la suma de amount de actual_expenses para la combinacion
    (year, month, cost_center_id, expense_concept_id).
    No incluye gastos planificados, desviacion ni porcentaje; esos calculos
    corresponden a tareas posteriores.
    """

    year: int
    month: int
    cost_center_id: int
    cost_center_code: str | None
    cost_center_name: str | None
    expense_concept_id: int
    expense_concept_code: str | None
    expense_concept_name: str | None
    actual_amount: Decimal


class PlannedExpenseConsolidatedRead(BaseModel):
    """Consolidado de gastos planificados agrupado por periodo, centro y concepto.

    planned_amount es la suma de amount de planned_expenses para la combinacion
    (year, month, cost_center_id, expense_concept_id).
    No incluye gastos reales, desviacion ni porcentaje; esos calculos
    corresponden a tareas posteriores.
    """

    year: int
    month: int
    cost_center_id: int
    cost_center_code: str | None
    cost_center_name: str | None
    expense_concept_id: int
    expense_concept_code: str | None
    expense_concept_name: str | None
    planned_amount: Decimal
